tls/ttls.c

```diff
@@ -2203,8 +2204,14 @@ ttls_recv(void *tls_data, unsigned char *buf, size_t len, unsigned int *read)
 	}
 	WARN_ON_ONCE(!io->msglen);
 	delta = *read - parsed;
-	if (delta == len)
+	if (delta == len) {
+		if (io->msgtype == TTLS_MSG_HANDSHAKE
+		    && ttls_hs_checksumable(tls))
+		{
+			ttls_update_checksum(tls, buf + delta - hh_len, hh_len);
+		}
```

Comment:
The fix is correct, but current code with checksumming and chunks is tricky and the fix makes it even more complicated. Here we care about handshake messages only, unencrypted since encryption takes place on CLIENT_FINISHED only and this state is not checksumable. So `ttls_parse_record_hdr()` returns `T_OK` only when it fully reads both the TLS and handshake headers and each of them can be split into several chunks.

If in current call we've read a chunk containing only tail of handshake header chunk, them we do calculate checksum above. We also go here with zero `hh_len`, i.e. do not do any checksumming actually. This is unclear and adds extra call without any actual work.

If current chunk contains the full handshake header, then we're going to compute it's checksum later, when we read more data after the header. But, yes, if we have no more data, then we do not compute checksum and we need the fix.

So to make the fix more clear I'd move the checksumming with `*read - parsed == len` condition into the double if above. And extend the comment with words that we go here if we read a chunk fully containing handshake header and probably we reached end of data, so we need to compute the checksum now.